cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.md

```diff
@@ -91,33 +91,27 @@ For these cases, attempting to retrofit this pattern in existing applications re
-- Includes an additional field called ‘operation’ that would indicate the purpose of the operation for which you are including the CSRF token (may it be form tag/ajax call)
-
-Example:
-
-```text
-oneclickpurchase (or) buy/asin=SDFH&category=2&quantity=3)
-```
-
-**Note:** Fields mentioned in encryption token pattern (user's ID, a timestamp value and a nonce) are included.
-
-The operation field helps in mitigating the fact that the hash function generates the same value irrespective of multiple iterations (unlike strong encryption functions that generate different values when they are encrypted each time). So, it would help in avoiding having repeated token values across your application. Nonce field serves the same purpose as in encrypted token pattern (i.e., to avoid rare collisions due to weak cryptographic functions and acts as a defense-in-depth measure).
-
-Generate the token using HMAC including all four fields mentioned previously (user's ID, a timestamp value, nonce, and operation) and then include it in hidden fields for form tags, headers/parameters for ajax calls. Once you receive the HMAC from the client in the requests, re-generate HMAC with the same fields that you used to generate it, and then verify that the HMAC you re-generated matches the HMAC received from the client. If it does, it is a legitimate user request and if it does not, flag it as a CSRF intrusion and alert your incident response teams. Because an attacker has no visibility into the key used for generating the hash fields used in generating it, there is no way for them to re-generate it to use in forged request.
-
+HMAC Based Token Pattern mitigation is also achieved without maintaining any state at the server. [HMAC](https://en.wikipedia.org/wiki/HMAC) based CSRF protection works similar to encryption based CSRF protection with a couple of minor differences
+1. Use strong HMAC function (SHA256 or stronger algorithms are recommended) instead of an encryption function to generate the token
+2. Token is HMAC+timestamp
+
+Below are the steps for the proper implementation of the HMAC based CSRF protection:
+1. Generate the token
+   Using key K, generate HMAC(usersessionID+timestamp) and append the same timestamp value to it which results in your CSRF token.
+2. Include the token (i.e., HMAC+timestamp) in a hidden field for forms and in the request-header field/request body parameter for AJAX requests 
+3. Validating the token 
+   When the request is received at the server, re-generate the token with same key K (parameters are sessionID from the request and timestamp in the received token). If the HMAC in the received token and the one generated in this step match, verify if timestamp received is less than defined token expiry time. If both of them are success, then request is treated as legitimate and can be allowed. If not, block the request and log the attack for incident response purposes.
```

Comment:
Same about here about the newline comment.
`are successful, then the request ...`